File: intrinsic_core/tools/sweep_parameters.cpp
// Finds the usable band of the discretiser's resolution.
//
// Too coarse and everything collapses into one bin (0 bits everywhere); too
// fine and nothing collapses at all (saturated everywhere). A usable
// resolution sits in the middle, away from both. Run this before trusting any
// map.
#include <cstdio>
#include <vector>

#include "intrinsic_core/arm2d.hpp"
#include "intrinsic_core/discretise.hpp"
#include "intrinsic_core/empowerment.hpp"

using namespace intrinsic_core;  // NOLINT(build/namespaces)

int main()
{
  const PlanarArm arm;
  const Eigen::Vector2d puck(0.30, 0.25);

  auto state_at = [&](double x, double y) {
      const auto ik = arm.inverse_kinematics(x, y);
      Eigen::VectorXd s(4);
      s << (*ik)[0], (*ik)[1], puck[0], puck[1];
      return s;
    };
  const Eigen::VectorXd near = state_at(0.30, 0.25);
  const Eigen::VectorXd far = state_at(0.30, 0.42);

  std::printf("%12s %10s %10s %10s\n", "resolution", "near", "far", "contrast");
  for (double r : {100.0, 1.0, 0.20, 0.10, 0.05, 0.025, 0.01, 0.005, 1e-9}) {
    EmpowermentConfig cfg;
    cfg.n_steps = 3;
    cfg.seed = 0;
    const EmpowermentEstimator est(
      [&arm](const Eigen::VectorXd & s, const std::vector<int> & a) {
        return arm.rollout(s, a);
      },
      9, grid_discretiser(r), cfg);
    const double n = est.estimate(near);
    const double f = est.estimate(far);
    std::printf("%12.6f %10.3f %10.3f %10.3f\n", r, n, f, n - f);
  }
  return 0;
}

File: intrinsic_core/scripts/sweep_parameters.py
#!/usr/bin/env python3
"""
Find the band where empowerment actually has structure.

The paper advertises empowerment as parameter-free. On a continuous system
it is not. Two parameters decide whether the map means anything, and both
have a failure mode on each side:

    outcome_resolution   too coarse -> every outcome in one bin -> 0 bits
                         too fine   -> no outcomes merge -> flat and high

    horizon              too short  -> nothing reachable -> flat and low
                         too long   -> everything reachable -> flat again
                                       (sec 4.5.5, the Greek gods problem)

This script measures the *contrast* -- empowerment at the object minus
empowerment away from it -- across both. Contrast, not absolute value, is
what you need: a map can have plenty of bits everywhere and still tell you
nothing.

Run this before trusting any result, and again whenever you change the arm.

    python3 scripts/sweep_parameters.py
"""

from __future__ import annotations

import sys
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from intrinsic_core import (  # noqa: E402
    EmpowermentConfig,
    EmpowermentEstimator,
    GridDiscretiser,
    PlanarArm,
)

PUCK = np.array([0.30, 0.25])
AT_OBJECT = (0.30, 0.25)
AWAY = (0.30, 0.42)


def _state(arm, xy):
    ik = arm.inverse_kinematics(*xy)
    return None if ik is None else np.array([ik[0], ik[1], *PUCK])


def contrast(arm, resolution: float, horizon: int) -> tuple[float, float, float]:
    est = EmpowermentEstimator(
        arm.rollout, 9, GridDiscretiser(resolution),
        EmpowermentConfig(n_steps=horizon, n_sequences=600, seed=0),
    )
    near = est.estimate(_state(arm, AT_OBJECT))
    far = est.estimate(_state(arm, AWAY))
    return near, far, near - far


def _verdict(near: float, delta: float, ceiling: float) -> str:
    if near < 0.1:
        return "DEAD  (all outcomes in one bin)"
    if near > ceiling - 0.15:
        return "SATURATED  (nothing merges)"
    if delta < 0.25:
        return "flat  (no contrast)"
    return "usable"


def main() -> int:
    arm = PlanarArm()

    print("\nRESOLUTION SWEEP  (horizon = 3)")
    print(f"{'res (m)':>10} {'at object':>11} {'away':>8} {'contrast':>10}   verdict")
    print("-" * 62)
    ceiling = 7.88  # measured degeneracy limit for this arm, see tests
    for r in [0.001, 0.005, 0.01, 0.02, 0.025, 0.04, 0.06, 0.10, 0.20, 1.0]:
        near, far, delta = contrast(arm, r, 3)
        print(f"{r:>10.3f} {near:>11.3f} {far:>8.3f} {delta:>+10.3f}   "
              f"{_verdict(near, delta, ceiling)}")

    print("\nHORIZON SWEEP  (resolution = 0.025 m)")
    print(f"{'steps':>10} {'at object':>11} {'away':>8} {'contrast':>10}   verdict")
    print("-" * 62)
    for n in [1, 2, 3, 4, 5]:
        near, far, delta = contrast(arm, 0.025, n)
        print(f"{n:>10d} {near:>11.3f} {far:>8.3f} {delta:>+10.3f}   "
              f"{_verdict(near, delta, ceiling)}")

    print("\nPick a resolution and horizon from the middle of the usable band,")
    print("not the edge. The edges are where the artefacts live.\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
